src/transforms/lua/v2/mod.rs

```diff
@@ -82,47 +92,67 @@ impl Lua {
                 package.set("path", paths)?;
             }
 
-            let func = ctx.load(&source).into_function()?;
-            ctx.set_named_registry_value("vector_func", func)?;
+            if let Some(hooks_init) = &config.hooks.init {
+                let hooks_init: rlua::Function<'_> =
+                    ctx.load(hooks_init).eval().context(InvalidHooksInit)?;
+                hooks_init.call(()).context(RuntimeErrorHooksInit)?;
+            }
+
+            let hooks_process: rlua::Function<'_> = ctx
+                .load(&config.hooks.process)
+                .eval()
+                .context(InvalidHooksProcess)?;
+            ctx.set_named_registry_value("hooks_process", hooks_process)?;
+
             Ok(())
-        })
-        .context(InvalidLua)?;
+        })?;
 
         Ok(Self {
             lua,
             invocations_after_gc: 0,
         })
     }
 
-    fn process(&mut self, event: Event) -> Result<Option<Event>, rlua::Error> {
-        let result = self.lua.context(|ctx| {
-            let globals = ctx.globals();
-
-            globals.set("event", event)?;
-
-            let func = ctx.named_registry_value::<_, rlua::Function<'_>>("vector_func")?;
-            func.call(())?;
-            globals.get::<_, Option<Event>>("event")
+    fn process(&mut self, event: Event, output: &mut Vec<Event>) -> Result<(), rlua::Error> {
+        let result = self.lua.context(|ctx: rlua::Context<'_>| {
+            ctx.scope(|scope| {
+                let emit = scope.create_function_mut(|_, event: Event| {
+                    output.push(event);
+                    Ok(())
+                })?;
+                let process = ctx.named_registry_value::<_, rlua::Function<'_>>("hooks_process")?;
+
+                process.call((event, emit))
+            })
         });
+
         self.invocations_after_gc += 1;
         if self.invocations_after_gc % GC_INTERVAL == 0 {
             self.lua.gc_collect()?;
             self.invocations_after_gc = 0;
         }
         result
     }
+
+    // Used only in tests.
```

Comment:
Can we `#[cfg(test)]` this?

Reply:
These are part of the [`Transform`](https://github.com/timberio/vector/blob/a73222e591eabf3e95a0467712bcdaf2855fae99/src/transforms/mod.rs#L60-L66) trait, so it is necessary to provide at least some implementation for them.